### ISSUE TITLE
Problem getting CDN to work

### ISSUE THREAD

**Issue by thalesfc** [dnouri/nolearn#81]
Hi guys, I want to configure a small Deep Belief Network to classify some data. 

First, I configured a DBN that only uses back propagation to train:
```
dbn = DBN(
    layer_sizes=[-1, 100, -1],
    learn_rates=0.3,
    learn_rate_decays=0.9,
    epochs=1,
    verbose=1
)
dbn.fit(trainX, trainY)
predicted = dbn.predict(testX)
print "Accuracy:", accuracy_score(testY, predicted)
print "Classification report:"
print classification_report(testY, predicted)
```
this code is working and produces the following output:
```
In [44]: %run dbn.py
$ Loading dataset=data/arrhythmia.data, with DEBUG=False
[DBN] fitting X.shape=(387, 183)
[DBN] layers [183, 100, 2]
[DBN] Fine-tune...
100%
Epoch 1:
  loss 0.673116505146
  err  0.4375
  (0:00:00)
Accuracy: 0.604651162791
Classification report:
             precision    recall  f1-score   support

        0.0       0.60      1.00      0.75        26
        1.0       0.00      0.00      0.00        17

avg / total       0.37      0.60      0.46        43
``` 

Next, I decided to include a CDN pre-training. 
```
dbn = DBN(
    ...
    epochs_pretrain=1,
    ...
)
```

But the code is **not working**, producing the following output:
```
In [45]: %run dbn.py
$ Loading dataset=data/arrhythmia.data, with DEBUG=False
[DBN] fitting X.shape=(387, 183)
[DBN] layers [183, 100, 2]
[DBN] Pre-train layer 1...
  Epoch 1: err nan
  (0:00:00)
[DBN] Pre-train layer 2...
  Epoch 1: err nan
  (0:00:00)
[DBN] Fine-tune...
100%
Epoch 1:
  loss nan
  err  0.408854166667
  (0:00:00)
Accuracy: 0.581395348837
Classification report:
             precision    recall  f1-score   support

        0.0       0.58      1.00      0.74        25
        1.0       0.00      0.00      0.00        18

avg / total       0.34      0.58      0.43        43
```

I think I am missing some parameters, does anyone have an idea about what is wrong?

**Comment by bmilde** [comment]
"Epoch 1: err nan" is your problem and suggests that your learning rate (0.03) is too high for the problem. Try 0.01 or 0.001.

Other than that, I'd suggest that you use the lasagne wrapper, as the DBN class is deprecated. So that would mean something like:

    nn = NeuralNet(
        layers=[
            ('input', InputLayer),
            ('hidden1', DenseLayer),
            ('dropout1', DropoutLayer),
            ('hidden2', DenseLayer),
            ('dropout2', DropoutLayer),
            ('output', DenseLayer),
            ],
        input_shape=(None, 183),
        output_num_units=2,
        output_nonlinearity=softmax,
        hidden1_num_units=100,
        hidden2_num_units=100,

        update=nesterov_momentum,
        update_learning_rate=0.01,
        update_momentum=0.9,

        max_epochs=500,
        )

**Comment by thalesfc** [comment]
Thank you @bmilde.

Another doubt, in your example of the lasagne wrapper, is it training using CDN? If not, how to adapt it to do so?

Thanks :)

**Comment by dnouri** [comment]
You're right that [Lasagne doesn't support pre-training](https://groups.google.com/d/topic/lasagne-users/ywDIOUVIsJQ/discussion).  But chances are that normal supervised training, or an autoencoder will work for you.  Autoencoders can be built pretty easily with Lasagne (and nolearn).
